common/cpp/src/google_smart_card_common/value_debug_dumping.cc

```diff
@@ -0,0 +1,124 @@
+// Copyright 2020 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <google_smart_card_common/value_debug_dumping.h>
+
+#include <string>
+
+#include <google_smart_card_common/logging/hex_dumping.h>
+#include <google_smart_card_common/logging/logging.h>
+#include <google_smart_card_common/value.h>
+
+namespace google_smart_card {
+
+namespace {
+
+std::string GetValueTypeTitle(const Value& value) {
+  switch (value.type()) {
+    case Value::Type::kNull:
+      return Value::kNullTypeTitle;
+    case Value::Type::kBoolean:
+      return Value::kBooleanTypeTitle;
+    case Value::Type::kInteger:
+      return Value::kIntegerTypeTitle;
+    case Value::Type::kFloat:
+      return Value::kFloatTypeTitle;
+    case Value::Type::kString:
+      return Value::kStringTypeTitle;
+    case Value::Type::kBinary:
+      return Value::kBinaryTypeTitle;
+    case Value::Type::kDictionary:
+      return Value::kDictionaryTypeTitle;
+    case Value::Type::kArray:
+      return Value::kArrayTypeTitle;
+  }
+  GOOGLE_SMART_CARD_NOTREACHED;
+}
+
+std::string DebugDumpBoolean(bool value) { return value ? "true" : "false"; }
+
+std::string DebugDumpString(const std::string& value) {
+  return '"' + value + '"';
+}
+
+std::string DebugDumpArray(const Value::ArrayStorage& array_value) {
+  std::string result = "[";
+  for (size_t index = 0; index < array_value.size(); ++index) {
+    if (index > 0) result += ", ";
+    result += DebugDumpValueFull(*array_value[index]);
+  }
+  result += "]";
+  return result;
+}
+
+std::string DebugDumpDictionary(
+    const Value::DictionaryStorage& dictionary_value) {
+  std::string result = "{";
+  bool is_first_item = true;
+  for (const auto& item : dictionary_value) {
+    const std::string& item_key = item.first;
+    const Value& item_value = *item.second;
+    if (!is_first_item) result += ", ";
+    is_first_item = false;
+    result += DebugDumpString(item_key);
+    result += ": ";
+    result += DebugDumpValueFull(item_value);
+  }
+  result += "}";
+  return result;
+}
+
+std::string DebugDumpBinary(const Value::BinaryStorage& binary_value) {
+  std::string result = std::string(Value::kBinaryTypeTitle) + "[";
```

Comment:
nit: This is completely not needed (and more verbose), feel free to drop it.

I see you are **only** prefixing the Value::BinaryStorage with its type string `std::string(Value::kBinaryTypeTitle)` (I presume this is because you want to differentiate between Array and Binary types, no?). Would it be useful to add the type before each DebugDump*, as well? for example, in `DebugDumpBoolean(bool value) { return std::string(Value::kBooleanTypeTitle) + ": " + (value ? "true" : "false"); }`

Reply:
Yeah, I was trying to keep these log strings reasonably short and clean. Things like `true` can be undoubtedly interpreted by the human as "boolean with the true value", so we don't really need the type information there. I understand your idea about being more consistent, but would prefer keeping the more compact format with one special case for binary data.
But I've added a comment about this into the code.